### PR TITLE
[WIP][HELP-REQ] Implementing `idxmin` and `idxmax` for `Series` and `DataFrame`

### LINKED ISSUE #2188: [FEA] Implement pandas idxmax and idxmin method
**Is your feature request related to a problem? Please describe.**

The `idxmax` function from pandas doesn't appear to be implemented on cudf. For my code, I particularly need this implemented on groupby objects, but it exists on pandas dataframe and series objects as well.

### PR DESCRIPTION
**Pull request by millerhooks**
## Overview
I have stubbed out the functions for `idxmin/max` for the `pandas` matching `DataFrame` and `Series` `idxmin` and `idxmax`  functions. I have also set up a complete test suite. I have removed my `libcudf` `C++` experiments and created this PR to get some dialog and guidance on implementing the `C++` API.

I am fairly new to `libcudf` development and would enjoy some direction on where to place the operations as well as anything that may not be obvious in implementation. Does `idxmin/max` fit under any existing `cudf` operation namespace? I did not see anywhere it would fit easily, but it doesn't feel significant enough to make it's own space.

## `pandas.Series.idxmin/idxmax` Operations 
For ease of reference to aid discussion, here are the locations of the operations in `pandas`.
```
       skipna = nv.validate_argmax_with_skipna(skipna, args, kwargs)
        i = nanops.nanargmax(com.values_from_object(self), skipna=skipna)
        if i == -1:
            return np.nan
        return self.index[i]
```
`pd` validate the `argmax` with `skipna` and get the value of that object as the index location for it's value. `nanargmax` will return the appropriate location based off applied `mask` and `skipna` while managing and filling `dtypes` appropriately. Finally, Returning `np.nan` if the selected object value is -1 and returning an index for the value if not.

### Code Links
- https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/core/series.py#L2214-L2283
- https://github.com/pandas-dev/pandas/blob/62a87bf4a2af02a8d3bc271ad26e5994292b8e6a/pandas/compat/numpy/function.py#L97-L132
- https://github.com/pandas-dev/pandas/blob/62a87bf4a2af02a8d3bc271ad26e5994292b8e6a/pandas/core/nanops.py#L856-L883
- https://github.com/pandas-dev/pandas/blob/62a87bf4a2af02a8d3bc271ad26e5994292b8e6a/pandas/compat/numpy/function.py#L97

## `pandas.DataFrame.idxmin/idxmax` Operations 
```
        axis = self._get_axis_number(axis)
        indices = nanops.nanargmin(self.values, axis=axis, skipna=skipna)
        index = self._get_axis(axis)
        result = [index[i] if i >= 0 else np.nan for i in indices]
        return Series(result, index=self._get_agg_axis(axis))
```

### Code Links
- https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/core/frame.py#L8047-L8082

closes #2188

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -3519,6 +3519,57 @@ def notnull(self):
...
+        # requirements established.
+        return self.to_pandas().idxmin(axis=axis, skipna=skipna)
+
+    def idxmax(self, axis=0, skipna=True):
+        """
+        Return index of first occurrence of maximum over requested axis.
+        NA/null values are excluded.
+
+        Parameters:
+            axis : {0 or ‘index’, 1 or ‘columns’}, default 0
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py]: Let's downscope for now and focus on only doing a column-wise `idxmax` for now. Let's update these docs accordingly to indicate axis=0 is not yet supported.

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -3519,6 +3519,57 @@ def notnull(self):
...
+    # `idxmin` and `idxmax` are stubbed here for writing expected
+    # tests. After feedback from the etl team on the libcudf API
+    # they will be replaced.
+    def idxmin(self, axis=0, skipna=True):
+        """
+        Return index of first occurrence of minimum over requested axis.
+        NA/null values are excluded.
+
+        Parameters:
+            axis : {0 or ‘index’, 1 or ‘columns’}, default 0
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py]: This one too

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/core/dataframe.py]: `axis=0` would require all columns to be the same dtype, and we'd need a implementation that implicitly transposes the table as if it were a matrix. Doable, but an entirely different scenario.

**Comment by cwharris** [comment]
For the libcudf api, you may be able to use `cudf::experimental::sorted_order`, then grab the first/last element value from the result. That would be your min/max idx, which you could then use to look up the label you need. It's possible no libcudf changes would be necessary for this.

**Comment by kkraus14** [comment]
> Since the memory space of sorted_order is O(N)

Isn't in `O(N log N)` in the worst case, but `O(N)` almost always because it's a radix sort for non-string types? Regardless, it feels like we should be able to be more efficient than sorting here.

**Comment by cwharris** [comment]
how does this look?
```
size_type
reduce_index(cudf::column_view const&
             cudf::experimental::reduction_op op);
```
(I think `rmm::mr::memory_resource` isn't necessary since it only returns `size_type`).

Would be defined behavior only for `reduction_op::MIN` or `reduction_op::MAX`.
Would return `-1` if all elements are invalid.

**Comment by harrism** [comment]
This should not be implemented using a sort. Cost of sort is O(N) extra space and O(N log N) work.  This should be implemented using an argmin/argmax reduction. Cost of reduction is O(1) extra space and O(N) work.

libcudf has argmin/argmax `cudf::aggregation` type, but I don't see any gtests using it for column reduction (only gtests using it for groupby). So it may not be implemented yet for column reduction.

**Comment by harrism** [comment]
Changing branch to 0.14 since this doesn't seem to be progressing.

**Comment by cwharris** [comment]
@harrism Honestly. What was I thinking. :man_facepalming:

**Comment by devavret** [comment]
@karthikeyann is currently developing `ARGMIN` and `ARGMAX` for column level.

Karthikeyan, you might want to think about adding the ability to do `ARGMIN/ARGMAX` on a table level for `DataFrame.idxmin() / .idxmax()`

**Comment by kkraus14** [comment]
Closing as this is stale.

**Pull request closed without merging**